### PR TITLE
sql: add a SHOW TENANTS command

### LINKED ISSUE #92641: sql: implement `SHOW TENANTS`
We already have `SHOW TENANT <name>`, it is useful to also have a flavor that shows all tenants instead of a specific one.

Epic: CRDB-18749

Jira issue: CRDB-21893

### PR DESCRIPTION
**Pull request by lidorcarmel**
This command prints all available tenants and their status.
To get the status for replication (c2c) tenants we need to get the job info.

For example:
```
    root@127.0.0.1:26257/defaultdb> show tenants;
      id |   name   |                               status
    -----+----------+----------------------------------------------------------------------
       1 | system   | ACTIVE
       2 | src      | ACTIVE
       3 | dest     | REPLICATION PAUSED
       4 | dest2    | REPLICATION UNKNOWN (succeeded)
       6 | dest3    | REPLICATION UNKNOWN (job with ID 819888096954253313 does not exist)
      11 | dest4    | ACTIVE
      12 | dest5    | REPLICATING
      13 | dest6    | INITIALIZING REPLICATION
    (13 rows)

    Time: 67ms total (execution 67ms / network 0ms)
```

Fixes: #92641

Epic: CRDB-18749

Release note: None

### CODE REVIEW

File: pkg/ccl/streamingccl/streamingest/alter_replication_job.go [hunk 1]
```diff
@@ -161,7 +161,7 @@ func alterReplicationJobHook(
 			return err
 		}
 		if tenInfo.TenantReplicationJobID == 0 {
-			return errors.Newf("tenant %q does not have an active replication job", tenantName)
+			return errors.Newf("tenant %q is not a replicated tenant", tenantName)
```
Comment (adityamaru) [inline pkg/ccl/streamingccl/streamingest/alter_replication_job.go:164]: hmm is this change necessary? Tenant foo could have been created as a result of c2c replication, so it could be confusing to say that it is not a replicated tenant if the job ID is 0. IMO it is sufficient to say that there is no active replication job associated with this tenant and so the replication information you are trying to access is unavailable.

File: pkg/sql/tenant.go [hunk 2]
```diff
@@ -202,6 +202,33 @@ func CreateTenantRecord(
...
+		ctx, "get-tenant-names", txn,
+		`SELECT name FROM system.tenants`,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	tenants := make([]roachpb.TenantName, 0, len(rows))
+	for _, tenant := range rows {
+		// If the tenants is in the DROP state the then name is nil.
```
Comment (adityamaru) [inline pkg/sql/tenant.go:222]: I believe tenant name can also be zero if it was created prior to V23_1TenantNames, or using the `create_tenant(id)` builtin.

File: pkg/sql/tenant.go [hunk 3]
```diff
@@ -202,6 +202,33 @@ func CreateTenantRecord(
 	)
 }
 
+func GetAllTenantNames(
```
Comment (adityamaru) [inline pkg/sql/tenant.go:205]: nit: linter might scream at you to add a comment to this exported method

File: pkg/sql/show_tenant.go [hunk 4]
```diff
@@ -124,89 +152,103 @@ func (n *showTenantNode) initReplicationStats(params runParams, job *jobs.Job) e
...
-	switch n.tenantInfo.State {
+	switch values.tenantInfo.State {
 	case descpb.TenantInfo_ADD:
 		// There is a replication job, we need to get the job info and the
 		// replication stats in order to generate the exact tenant status.
 		registry := params.p.execCfg.JobRegistry
 		job, err := registry.LoadJobWithTxn(params.ctx, jobId, params.p.Txn())
 		if err != nil {
-			return err
+			values.tenantStatus = tenantStatus(fmt.Sprintf(string(replicationUnknownFormat), err))
```
Comment (adityamaru) [inline pkg/sql/show_tenant.go:192]: It makes me a little nervous that we aren't failing loudly enough here. Should we log `ERROR: %s` instead of `REPLICATION UNKNOWN`, here and below?

File: pkg/sql/show_tenant.go [hunk 5]
```diff
@@ -124,89 +152,103 @@ func (n *showTenantNode) initReplicationStats(params runParams, job *jobs.Job) e
...
 		default:
-			n.tenantStatus = tenantStatus(fmt.Sprintf(string(replicationUnknownFormat), job.Status()))
+			values.tenantStatus = tenantStatus(fmt.Sprintf(string(replicationUnknownFormat), job.Status()))
 		}
 	case descpb.TenantInfo_ACTIVE, descpb.TenantInfo_DROP:
-		n.tenantStatus = tenantStatus(n.tenantInfo.State.String())
+		values.tenantStatus = tenantStatus(values.tenantInfo.State.String())
 	default:
-		return errors.Newf("unknown tenant status %s", n.tenantInfo.State)
+		status := fmt.Sprintf("tenant status: %s", values.tenantInfo.State)
```
Comment (adityamaru) [inline pkg/sql/show_tenant.go:227]: I prefer if we leave this an error. If we see a tenant in a state other than `ADD`, `DROP`, `ACTIVE` then it means someone has added a state and not considered all the places it effects, so I'd rather it fail tests loudly. wdyt?

**Review by adityamaru — CHANGES_REQUESTED** [review]

File: pkg/sql/show_tenant.go (hunk 4 above)
Comment (stevendanna) [reply pkg/sql/show_tenant.go:192]: Agree. If we have a non-zero jobID, that job ID should have a status we should be able to look up. This presents a non-trivial error I think.

File: pkg/sql/show_tenant.go [hunk 6]
```diff
@@ -93,100 +110,145 @@ func (n *showTenantNode) getTenantName(params runParams) (roachpb.TenantName, er
...
+			protectedTimestamp = record.Timestamp
+		}
 	}
+	return stats, &protectedTimestamp, nil
+}
+
+func (n *showTenantNode) getTenantValues(
+	params runParams, tenantName roachpb.TenantName,
+) (*tenantValues, error) {
 	tenantRecord, err := GetTenantRecordByName(params.ctx, params.p.execCfg, params.p.Txn(), tenantName)
```
Comment (stevendanna) [inline pkg/sql/show_tenant.go:169]: As a heads up, there is another PR in flight to make most of these commands work with either IDs _or_ names. Not sure there is anything we want to do here other than being willing to help resolve the conflict when it occurs :D.

File: pkg/sql/show_tenant.go [hunk 7]
```diff
@@ -93,100 +110,145 @@ func (n *showTenantNode) getTenantName(params runParams) (roachpb.TenantName, er
 }
 
 func (n *showTenantNode) startExec(params runParams) error {
-	tenantName, err := n.getTenantName(params)
+	if n.all {
+		names, err := GetAllTenantNames(params.ctx, params.p.execCfg, params.p.Txn())
```
Comment (stevendanna) [inline pkg/sql/show_tenant.go:114]: I'm wondering, instead of passing this list of names around. Could we do more of this in sql?  That is, we are essentially pretty-printing the results of `SELECT id,  FROM system.tenants` and annotating it with job details. I wonder if we could either: (1) use an iterator over a query we construct to select the correct nodes or (2) implement this as a delegated statement instead.

I'm OK doing this as a follow-up, just a _little_ suspect of anywhere that we require passing around complete sets of user-generatable data.

**Review by lidorcarmel — COMMENTED** [review]
unfortunately too many changes :(
PTAL
mostly - this pr now gets the list of ids, not names, which means we also show dropped tenants without a name! (I thought of putting the droppedName there but then the name would not be unique which is not fun). because of that this pr is sorting the list by tenant id (probably better than by name?).
added some tests.
thanks!!

**Review by stevendanna — COMMENTED** [review]
Overall, this seems reasonable to me. I think a few new logic tests would be great though.

**Review by lidorcarmel — COMMENTED** [review]
thanks! done.

**Review by adityamaru — APPROVED** [review]

**Comment by lidorcarmel** [comment]
thanks both for the review!
bors r+
